File: javalin/src/main/java/io/javalin/http/sse/Emitter.java
package io.javalin.http.sse;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import javax.servlet.AsyncContext;
import javax.servlet.ServletOutputStream;

class Emitter {

    private static final String newline = "\n";

    private final AsyncContext asyncContext;
    private ServletOutputStream output;
    private boolean close = false;

    public Emitter(AsyncContext asyncContext) {
        this.asyncContext = asyncContext;
        try {
            this.output = asyncContext.getResponse().getOutputStream();
        } catch (IOException e) {
            close = true;
        }
    }

    public void emit(@NotNull String event,@NotNull String data, String id) {
        synchronized(this) {
            try {
                StringBuilder sb = new StringBuilder();
                if (id != null) {
                    sb.append("id: ").append(id).append(newline);
                }
                sb.append("event: ").append(event).append(newline);
                for (String line : data.split("\\r?\\n")) {
                    sb.append("data: ").append(line).append(newline);
                }
                sb.append(newline).append(newline);
                output.print(sb.toString());
                asyncContext.getResponse().flushBuffer();
            } catch (IOException e) {
                close = true;
            }
        }
    }

    public boolean isClose(){
      return close;
    }

}

File: javalin/src/main/java/io/javalin/http/sse/Emitter.kt
package io.javalin.http.sse

import jakarta.servlet.http.HttpServletResponse
import java.io.IOException
import java.io.InputStream
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

const val COMMENT_PREFIX = ":"
const val NEW_LINE = "\n"

class Emitter(private var response: HttpServletResponse) {

    private val lock = ReentrantLock()

    var closed = false
        private set

    fun emit(event: String?, data: InputStream, id: String?) = lock.withLock {
        try {
            // Strip CR/LF from event and id so attacker-controlled values
            // can't inject extra SSE frames — see #2579.
            val sanitizedId = id?.let(::stripCrLf)
            val sanitizedEvent = event?.let(::stripCrLf)
            if (sanitizedId != null) {
                write("id: $sanitizedId$NEW_LINE")
            }
            if (sanitizedEvent != null) {
                write("event: $sanitizedEvent$NEW_LINE")
            }
            data.buffered().reader().useLines {
                it.forEach { line -> write("data: $line$NEW_LINE") }
            }

            write(NEW_LINE)
            response.flushBuffer()
        } catch (ignored: IOException) {
            closed = true
        }
    }

    private fun stripCrLf(value: String): String =
        value.replace("\r", "").replace("\n", "")

    fun emit(comment: String) =
        try {
            comment.split(NEW_LINE).forEach {
                write("$COMMENT_PREFIX $it$NEW_LINE")
            }
            response.flushBuffer()
        } catch (ignored: IOException) {
            closed = true
        }

    private fun write(value: String) =
        response.outputStream.print(value)

}
